components/suggest/src/util.rs

```diff
@@ -0,0 +1,1517 @@
+    filter_map_chunks_recurse(
+        &words,
+        &normalized_query,
+        &mut vec![],
+        0,
+        max_chunk_size,
+        &f,
+    )
+}
+
+/// `remaining_words` is the slice of remaining words in the query string at
+/// this step. `remaining_query` is the remaining slice of the query string at
+/// this step.
+///
+/// `path` is the sequence of values returned by the filter-map function so far
+/// at this step.
+///
+/// `chunk_index` is the word-based index in the query string at this step.
+fn filter_map_chunks_recurse<T: Clone>(
+    remaining_words: &[&str],
+    remaining_query: &str,
+    path: &mut Vec<T>,
+    chunk_index: usize,
+    max_chunk_size: usize,
+    f: &impl Fn(&str, usize, &[T]) -> Result<Option<Vec<T>>>,
+) -> Result<Vec<Vec<T>>> {
+    // Filtered-in (non-pruned) paths that will be returned from this step of
+    // the traversal.
+    let mut this_step_paths: Vec<Vec<T>> = vec![];
+
+    for chunk_size in 1..=max_chunk_size {
+        if remaining_words.len() < chunk_size {
+            // `chunk_size` and the later chunk sizes in this for-loop are too
+            // big to visit the remaining words. We already visited them earlier
+            // in the loop when the chunk size was small enough.
+            break;
+        }
+
+        // Get the current chunk within the remaining query. Its char length is
+        // the sum of the lengths of the words in the chunk + `chunk_size - 1`
+        // spaces between the words.
+        let chunk_char_len = remaining_words[..chunk_size]
+            .iter()
+            .fold(chunk_size - 1, |memo, w| memo + w.len());
+        let chunk = &remaining_query[..chunk_char_len];
+
+        // Call the mapper function.
+        if let Some(mapped_values) = f(chunk, chunk_index, &path[..])? {
+            for value in mapped_values {
+                if chunk_size == remaining_words.len() {
+                    // This is the final chunk in the path. Stop recursing.
+                    this_step_paths.push(vec![value.clone()]);
+                } else {
+                    // Recurse. Note that the new `remaining_words` slice won't
+                    // be empty because if it were, `chunk_size` would equal the
+                    // remaining word count, which is if-branch condition.
+                    path.push(value.clone());
+                    let subtree_paths = filter_map_chunks_recurse(
+                        &remaining_words[chunk_size..],
+                        &remaining_query[(chunk_char_len + 1)..],
+                        path,
+                        chunk_index + chunk_size,
+                        max_chunk_size,
+                        f,
+                    )?;
+                    path.pop();
+                    for mut p in subtree_paths {
+                        p.insert(0, value.clone());
+                        this_step_paths.push(p);
+                    }
+                }
```

Comment:
Suggestion: use `get()` to let Rust do the bounds check for you and remove the comment about it.
```
match remaining_words.get(chunk_size..) {
   None => {
      // final chunk
   }
   Some(remaining_words) => {
       // Recurse
   }
}
```

This could also be an `if let`.

Reply:
This doesn't quite work because at this point the chunk isn't ever out of bounds, due to the earlier `break` in the outer for-loop.